src/Microsoft.Azure.CosmosEventSourcing/Stores/DefaultEventStore.cs

```diff
@@ -40,6 +43,76 @@ internal class DefaultEventStore<TEventItem> :
             cancellationToken);
     }
 
+    public async ValueTask PersistAsync(
+        IAggregateRoot aggregateRoot,
+        CancellationToken cancellationToken = default)
+    {
+        List<PropertyInfo> partitionKeyProperties = aggregateRoot
+            .GetType()
+            .GetProperties()
+            .Where(x
+                => x.GetCustomAttributes()
+                    .Any(y =>
+                        y is EventItemPartitionKeyAttribute))
+            .ToList();
+
+        switch (partitionKeyProperties.Count)
+        {
+            case 0:
+                throw new InvalidOperationException(
+                    $"A {nameof(EventItemPartitionKeyAttribute)} must be present on a property in {aggregateRoot.GetType().Name}");
+            case > 1:
+                throw new InvalidOperationException(
+                    $"{nameof(EventItemPartitionKeyAttribute)} can not be present on multiple properties in {aggregateRoot.GetType().Name}");
+        }
+
+        Object partitionKey = partitionKeyProperties.Single().GetValue(aggregateRoot) ??
+                              throw new InvalidOperationException();
+
+        List<TEventItem?> events = aggregateRoot.NewEvents.Select(x =>
+        Activator.CreateInstance(
+            typeof(TEventItem),
+            x,
+            partitionKey) as TEventItem).ToList();
+
+        events.Add(Activator.CreateInstance(
+            typeof(TEventItem),
+            aggregateRoot.AtomicEvent,
+            partitionKey) as TEventItem);
+
+        if (events.Any(x => x == null))
+        {
+            throw new InvalidOperationException(
+                $"At least one of the {typeof(TEventItem).Name} could not be constructed");
+        }
+
+        await PersistAsync(events!, cancellationToken);
+    }
+
+    public async ValueTask PersistAsync(
+        IAggregateRoot aggregateRoot,
+        string partitionKeyValue,
+        CancellationToken cancellationToken = default)
+    {
+        List<TEventItem?> events = aggregateRoot.NewEvents.Select(x =>
```

Comment:
This code looks to be repeated can it be refactored into a private method and pass the partition key in?

Reply:
Good spot.

Reply:
Refactored